test/sim/abilities/protean.js

```diff
@@ -28,18 +32,90 @@ describe('Protean', function () {
+		], [
+			{species: "Wynaut", moves: ['sleeptalk']},
+		]]);
+
+		let kecleon = battle.p1.active[0];
+
+		// Fling with no item
+		battle.makeChoices('move fling', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Fling changed the user's type when it should not have.");
+
+		// Sucker Punch into status move
+		battle.makeChoices('move suckerpunch', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Sucker Punch changed the user's type when it should not have.");
+
+		// Steel Roller with no Terrain active
+		battle.makeChoices('move steelroller', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Steel Roller changed the user's type when it should not have.");
+
+		// Aura Wheel when user is not Morpeko
+		battle.makeChoices('move aurawheel', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Aura Wheel changed the user's type when it should not have.");
+
+		battle.makeChoices('switch 2', 'move sleeptalk');
+		kecleon = battle.p1.active[0];
+
+		// Burn Up when user is not Fire-type
+		battle.makeChoices('move burnup', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Burn Up changed the user's type when it should not have.");
+
+		// Aurora Veil when hail is not active
+		battle.makeChoices('move auroraveil', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Aurora Veil changed the user's type when it should not have.");
+
+		// Magnet Rise when user has Ingrain; note that Kecleon becomes Grass-type from Ingrain
+		battle.makeChoices('move ingrain', 'move sleeptalk');
+		battle.makeChoices('move magnetrise', 'move sleeptalk');
+		assert(kecleon.hasType('Grass'), "Magnet Rise changed the user's type when it should not have.");
+
+		// More examples: https://www.smogon.com/forums/threads/sword-shield-battle-mechanics-research.3655528/post-8548957
 	});
 
-	it('should not change the user\'s type when using Fling with no item', function () {
-		battle = common.createBattle();
-		battle.setPlayer('p1', {team: [{species: "Kecleon", ability: 'protean', moves: ['fling']}]});
-		battle.setPlayer('p2', {team: [{species: "Wobbuffet", moves: ['counter']}]});
-		battle.makeChoices('move fling', 'move counter');
-		assert(battle.p1.active[0].hasType('Normal'));
+	it(`should not change the user's type when abilities that fail earlier than Protean will activate`, function () {
+		battle = common.createBattle([[
+			{species: "Kecleon", ability: 'protean', moves: ['aquajet', 'mindblown']},
+		], [
+			{species: "Kyogre", ability: 'primordialsea', moves: ['sleeptalk']},
+			{species: "Groudon", ability: 'desolateland', moves: ['powder']},
+			{species: "Tsareena", ability: 'dazzling', moves: ['sleeptalk']},
+			{species: "Golduck", ability: 'damp', moves: ['sleeptalk']},
+		]]);
+
+		const kecleon = battle.p1.active[0];
+
+		// Primordial Sea
+		battle.makeChoices('move mindblown', 'move sleeptalk');
+		assert(kecleon.hasType('Normal'), "Primordial Sea changed the user's type when it should not have.");
```

Comment:
I also think these should probably say something more along the lines of "<ability> allowed the user's type to change" since Protean would be the one actually changing the user's type.

Reply:
I want to make it clear which effect is causing the move to fail. Otherwise, if it's 1 failing and they all have the same error come back, it'd be annoying to track down which of the 5 was wrong. I did a compromise of "Protean + thing".